src/include/duckdb/function/table_function.hpp

```diff
@@ -280,6 +281,8 @@ typedef unique_ptr<LocalTableFunctionState> (*table_function_init_local_t)(Execu
 typedef unique_ptr<BaseStatistics> (*table_statistics_t)(ClientContext &context, const FunctionData *bind_data,
                                                          column_t column_index);
 typedef void (*table_function_t)(ClientContext &context, TableFunctionInput &data, DataChunk &output);
+typedef SourceResultType (*async_table_function_t)(ClientContext &context, TableFunctionInput &data, DataChunk &output,
```

Comment:
As I've mentioned, I don't think this should directly take the `InterruptState`, performing the block should instead be a separate callback.

This is because there are scenarios where an operator is not allowed to block.
I think the logic here should be: return `SourceResultType::BLOCKED` to express intent to block, then the executor/global state of the operator is used to check whether the block is allowed to happen, when the block is allowed to happen, the callback to perform the block is issued, which receives the `InterruptState`